### PR TITLE
Vested Token implementation (Standard Token with vesting calendar)

### PR DESCRIPTION
**Pull request by izqui**
At [Aragon](https://aragon.one) we use Zeppelin's **Standard Token** as the base for our Stock contracts. One important feature in company stocks is the possibility to create a stock grant where you send some stock to a given entity, but that stock isn't fully theirs until a certain time passes, while they continue their relationship with the company, and it this were to change, have the ability for the company to get back non vested stock. 

In the current implementation, when a granter creates a grant, all tokens will be transferred to the grantee immediately. All non vested tokens at any given time will be locked and the grantee won't be allowed to transfer the tokens (`transfer(to, value)` has been re-implemented to check this). At any moment, granter can revoke the token grant and all non-vested tokens will be returned to her.

Creating a stock grant requires 3 parameters (start timestamp, cliff timestamp and vesting timestamp). From start to cliff, grantee won't be able to transfer any tokens from this grant. After cliff date, `tokens * (cliff - start) / (vesting - start)` tokens will unlock and from that moment until vesting date, grantee at timestamp `T` will have `tokens * (T - start) / (vesting - start)`tokens unlocked.

I have worked on some tests for GrantableToken, though they are not exhaustive yet given the fact that the implementation might change. Also I'm having trouble reliably implementing time delays to test token transferability, and some tests randomly fail sometimes (I think it is because of how TestRPC rounds milliseconds to seconds to pass to the EVM).

### CODE REVIEW

**Comment by maraoz** [comment]
Awesome contribution. Will review the code soon

**Comment by federicobond** [comment]
Looks really good! A couple of questions/suggestions:

 * Is there a reason for emphasizing the grant part as opposed to the vesting? At least from a cursory examination of the code, `VestingToken` or `VestedToken` seems to convey the meaning much more clearly for me.
 * I would replace the spelling of `transferrable` for `transferable` (single r), which is in much more widespread use today.
 * I would move the cliff calculation to a separate method, for easier testing and increased flexibility in vesting schemes.

**Comment by izqui** [comment]
@federicobond Thanks for the feedback, points 2 and 3 make a lot of sense and I have just modified it in this PR. 

Regarding naming I think that is a more abstract discussion, in my point of view the differentiable fact about this token is that rather than being only transferable it can be granted (and this grants can be revoked by granter). But if people are going to be confused for the name I'm cool with the name being changed to something like `VestingToken` or `VestableToken` (if that word even exists)

File: contracts/token/GrantableToken.sol [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    if (time < cliff) return 0;
+    if (time > vesting) return tokens;
+
+    uint256 cliffTokens = safeMul(tokens, (cliff - start) / (vesting - start));
+    vestedTokens = cliffTokens;
+
+    uint256 vestingTokens = safeSub(tokens, cliffTokens);
+
+    vestedTokens = safeAdd(vestedTokens, vestingTokens * (time - cliff) / (vesting - start));
+  }
```
Comment (maraoz) [inline contracts/token/GrantableToken.sol]: any reason you used safe math for some operations and not for others?

File: contracts/token/GrantableToken.sol [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+  function nonVestedTokens(TokenGrant grant, uint64 time) private constant returns (uint256) {
+    return safeSub(grant.value, vestedTokens(grant, time));
+  }
+
+  function lastTokenIsTransferrableEvent(address holder) constant public returns (uint64 date) {
```
Comment (maraoz) [inline contracts/token/GrantableToken.sol]: I suggest the name `lastTokenIsTransferableDate`
- Event is very semantically charged in Solidity.
- Transferrable => transferable

File: contracts/token/GrantableToken.sol [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    return super.transfer(_to, _value);
+  }
+
+  function max64(uint64 a, uint64 b) private constant returns (uint64) {
+    return a >= b ? a : b;
+  }
+
+  function min256(uint256 a, uint256 b) private constant returns (uint256) {
+    return a < b ? a : b;
+  }
```
Comment (maraoz) [inline contracts/token/GrantableToken.sol]: What do you think of moving these to SafeMath.sol?

File: contracts/token/GrantableToken.sol [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    uint64 cliff;
+    uint64 vesting;
+    uint64 start;
+  }
+
+  mapping (address => TokenGrant[]) public grants;
+
+  function grantTokens(address _to, uint256 _value) {
+    transfer(_to, _value);
+  }
```
Comment (izqui) [inline contracts/token/GrantableToken.sol]: Legacy from Aragon's codebase, can be removed

File: contracts/token/GrantableToken.sol (hunk 1 above)
Comment (izqui) [inline contracts/token/GrantableToken.sol]: I was just adding safeDivision to SafeMath and make all operations safe.

File: contracts/token/GrantableToken.sol (hunk 2 above)
Comment (izqui) [inline contracts/token/GrantableToken.sol]: Agree

File: contracts/token/GrantableToken.sol (hunk 3 above)
Comment (izqui) [inline contracts/token/GrantableToken.sol]: Makes sense

**Comment by izqui** [comment]
Everything discussed has been modified

**Comment by maraoz** [comment]
@izqui thanks for the modifications! I'm working on a fix to the failing test now

**Comment by maraoz** [comment]
@izqui Just opened a PR (https://github.com/AragonOne/zeppelin-solidity/pull/1) working towards fixing the failing test. The most important change is removing the busy waiting in the timer for an rpc call that moves the simulated node time forward. This should fix the random failing of tests. Feel free to merge this changes and try to fix the remaining failing tests, I haven't been able to do it :(

**Comment by izqui** [comment]
Tests are working for me now locally. Changed the base date to get the EVM time and also the order in which operations happen, multiplying before dividing, yay no floats in Solidity :(

**Comment by maraoz** [comment]
LGTM! Great contribution @izqui, thanks a lot for this!

**Pull request merged**